tests/test_pytorch/test_fsdp.py

```diff
@@ -0,0 +1,791 @@
+from torch.distributed.fsdp.wrap import size_based_auto_wrap_policy, wrap
+from torchmetrics import Accuracy
+
+if module_available("lightning"):
+    from lightning.fabric.utilities.imports import (
+        _TORCH_GREATER_EQUAL_2_0,
+        _TORCH_GREATER_EQUAL_2_1,
+    )
+    from lightning.pytorch import Trainer, seed_everything
+    from lightning.pytorch.demos import LightningTransformer
+    from lightning.pytorch.demos.boring_classes import BoringDataModule, BoringModel
+elif module_available("pytorch_lightning"):
+    from lightning_fabric.utilities.imports import (
+        _TORCH_GREATER_EQUAL_2_0,
+        _TORCH_GREATER_EQUAL_2_1,
+    )
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.demos import LightningTransformer
+    from pytorch_lightning.demos.boring_classes import BoringDataModule, BoringModel
+
+from lightning_habana.pytorch.accelerator import HPUAccelerator
+from lightning_habana.pytorch.plugins.fsdp_precision import HPUFSDPPrecision, HPUPrecisionPlugin
+from lightning_habana.pytorch.strategies import HPUDDPStrategy, HPUFSDPStrategy
+
+
+class TestFSDPModel(BoringModel):
+    def __init__(self):
+        super().__init__()
+        self.layer: Optional[nn.Module] = None
+
+    def _init_model(self) -> None:
+        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+
+    def configure_model(self) -> None:
+        if self.layer is None:
+            self._init_model()
+        # the model is already wrapped with FSDP: no need to wrap again!
+        if isinstance(self.layer, FullyShardedDataParallel):
+            return
+
+        self.layer = wrap(self.layer)
+
+    def configure_optimizers(self):
+        # There is some issue with SGD optimizer state in FSDP
+        return torch.optim.AdamW(self.layer.parameters(), lr=0.1)
+
+    def on_train_batch_start(self, batch, batch_idx):
+        assert batch.dtype == torch.float32
+
+    def on_train_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_test_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_validation_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def on_predict_batch_end(self, _, batch, batch_idx):
+        assert batch.dtype == torch.float32
+        self._assert_layer_fsdp_instance()
+
+    def _assert_layer_fsdp_instance(self) -> None:
+        assert isinstance(self.layer, FullyShardedDataParallel)
+        assert isinstance(self.trainer.strategy.precision_plugin, HPUFSDPPrecision)
+
+        if self.trainer.precision == "16-mixed":
+            param_dtype = None if not _TORCH_GREATER_EQUAL_2_0 else torch.float32
```

Comment:
we are already at 2.2. Do we need this check ?

Reply:
removed.